src/tiffvisitor_int.cpp

```diff
@@ -462,6 +463,77 @@ namespace Exiv2 {
         }
     } // TiffMetadataDecoder::decodeIptc
 
+    static const TagInfo* find(const TagInfo* pList,uint16_t tag)
+    {
+        while ( pList->tag_ != 0xffff && pList->tag_ != tag ) pList++;
+        return  pList->tag_ != 0xffff  ? pList : NULL;
+    }
+
+    void TiffDecoder::decodeCanonAFInfo(const TiffEntryBase* object) {
+        // report Exif.Canon.AFInfo as usual
+        TiffDecoder::decodeStdTiffEntry(object);
+        if ( !object->pValue()->count() || object->pValue()->typeId() != unsignedShort ) return; // no data!
+
+        // create vector of signedShorts from unsignedShorts in Exif.Canon.AFInfo
+        std::vector<int16_t>  ints;
+        std::vector<uint16_t> uint;
+        for (int i = 0; i < object->pValue()->count(); i++) {
+            ints.push_back((int16_t) object->pValue()->toLong(i));
+            uint.push_back((uint16_t) object->pValue()->toLong(i));
+        }
+        // Check this is AFInfo2 (ints[0] = bytes in object)
+        if ( ints[0] != object->pValue()->count()*2 ) return ;
+
+        std::string familyGroup(std::string("Exif.") + groupName(object->group()) + ".");
+
+        const int16_t nPoints = ints[2];
+        const int16_t nMasks  = (nPoints+15)/(sizeof(uint16_t) * 8);
+        int     nStart  = 0;
+
+        struct {
+        	uint16_t tag  ;
+        	uint16_t size ;
+        	bool     bSigned ;
+        } records[] = {
+        	{ 0x2600 , 1       , true  }, // AFInfoSize
+        	{ 0x2601 , 1       , true  }, // AFAreaMode
+        	{ 0x2602 , 1       , true  }, // AFNumPoints
+        	{ 0x2603 , 1       , true  }, // AFValidPoints
+        	{ 0x2604 , 1       , true  }, // AFCanonImageWidth
+        	{ 0x2605 , 1       , true  }, // AFCanonImageHeight
+        	{ 0x2606 , 1       , true  }, // AFImageWidth"
+        	{ 0x2607 , 1       , true  }, // AFImageHeight
+        	{ 0x2608 , nPoints , true  }, // AFAreaWidths
+        	{ 0x2609 , nPoints , true  }, // AFAreaHeights
+        	{ 0x260a , nPoints , true  }, // AFXPositions
+        	{ 0x260b , nPoints , true  }, // AFYPositions
+        	{ 0x260c , nMasks  , false }, // AFPointsInFocus
+        	{ 0x260d , nMasks  , false }, // AFPointsSelected
+        	{ 0x260e , 1       , true  }, // AFPrimaryPoint
```

Comment:
... and AFPrimaryPoint must have a length of `nMasks`.

Reply:
I agree with everything you've said.  Fixes submitted.